core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -1689,6 +1690,123 @@ RexNode simplifyAnd2ForUnknownAsFalse(List<RexNode> terms,
     }
   }
 
+  /**
+   * <p>Simplifies AND/OR condition that has common expressions,
+   * extract and eliminate/merge them as much as possible.</p>
+   * <ul>
+   * <li>(a OR b) AND (a OR b OR c OR d) =&gt; (a OR b)
+   * <li>(a OR b OR c OR d) AND (a OR b OR e OR f) =&gt; (a OR b) OR ((c OR d) AND (e OR f))
+
+   * <li>(a AND b) OR (a AND b AND c) =&gt; (a AND b)
+   * <li>(a AND b AND c AND d) OR (a AND b AND e AND f) =&gt;
+   * (a AND b) AND ((c AND d) OR (e AND f))
+   * </ul>
+   * <p> The difference between {@link #simplifyAnd} is that {@link #simplifyAnd} mainly
+   * simplifies expressions whose answer can be determined without evaluating both sides,
+   * like: FALSE AND (xxx) =&gt; FALSE</p>
+   */
+  public RexNode eliminateCommonExprInCondition(RexNode call) {
+    if (!(call instanceof RexCall)) {
+      return call;
+    }
+    // Simplify children recursively
+    // so that when we simplify parent node
+    // the child node should be simplified
+    List<RexNode> operands = ((RexCall) call).getOperands();
+    List<RexNode> simplifiedChildren = operands.stream()
+        .map(this::eliminateCommonExprInCondition)
+        .collect(Collectors.toList());
+
+    switch (call.getKind()) {
+    case AND: {
+      // Split the children to get the disjunctive predicates.
+      List<List<RexNode>> disjunctions = simplifiedChildren.stream()
+          .map(RelOptUtil::disjunctions)
+          .collect(Collectors.toList());
+
+      // Find the common predict among the list by retainAll.
+      List<RexNode> common = new ArrayList<>(disjunctions.get(0));
+      disjunctions.forEach(common::retainAll);
+
+      if (common.isEmpty()) {
+        // no common, return the original but with child simplified
+        return and(simplifiedChildren);
+      } else {
```

Comment:
Personally, I think avoiding else after return would make this method's code more clear.

Reply:
Sure~

Reply:
Hi @rubenada , seems to replace the "return" in switch makes the code longer😂

Reply:
you can take a look: https://github.com/apache/calcite/pull/2253/commits/65c35791fec7e1f9cec7b0ecd69f7f717404690b
cuz in the switch, we mean to break the branch, the return has this meaning.

Reply:
Maybe I did not explain myself clearly, I meant keeping the original code as it was, just removing the "else" instructions (which were not required after the "return"):
```
if (common.isEmpty())
  return and(simplifiedChildren);

disjunctions.forEach(p -> p.removeAll(common));
if (disjunctions.stream().anyMatch(List::isEmpty))
  return or(common);

List<RexNode> splice = disjunctions.stream().map(this::or).collect(Collectors.toList());
return or(or(common), and(splice));
```

Reply:
Sorry I misunderstood 😂

Reply:
Revised~